atlasdb-workload-server/src/main/java/com/palantir/atlasdb/workload/invariant/DurableWritesInvariant.java

```diff
@@ -0,0 +1,101 @@
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Consumer;
+import one.util.streamex.EntryStream;
+import one.util.streamex.StreamEx;
+
+public class DurableWritesInvariant implements Invariant {
+
+    private static final SafeLogger log = SafeLoggerFactory.get(DurableWritesInvariant.class);
+
+    private final Consumer<Map<TableWorkloadCell, MismatchedValue>> mismatchingCellsConsumer;
+    private final Consumer<Map<TableWorkloadCell, Integer>> undeletedCellsConsumer;
+
+    public DurableWritesInvariant() {
+        this(DurableWritesInvariant::logMismatchingCells, DurableWritesInvariant::logUndeletedCells);
+    }
+
+    @VisibleForTesting
+    DurableWritesInvariant(
+            Consumer<Map<TableWorkloadCell, MismatchedValue>> mismatchingCellsConsumer,
+            Consumer<Map<TableWorkloadCell, Integer>> undeletedCellsConsumer) {
+        this.mismatchingCellsConsumer = mismatchingCellsConsumer;
+        this.undeletedCellsConsumer = undeletedCellsConsumer;
+    }
+
+    @Override
+    public void accept(WorkflowHistory workflowHistory) {
+        InMemoryValidationStore expectedState = InMemoryValidationStore.create(workflowHistory.history());
+        mismatchingCellsConsumer.accept(
+                findCellsThatDoNotMatch(expectedState.values(), workflowHistory.transactionStore()));
+        undeletedCellsConsumer.accept(
+                findCellsThatAreNotDeleted(expectedState.deletedCells(), workflowHistory.transactionStore()));
+    }
+
+    private Map<TableWorkloadCell, MismatchedValue> findCellsThatDoNotMatch(
+            Map<TableWorkloadCell, Integer> expectedCells, ReadableTransactionStore storeToValidate) {
+        return EntryStream.of(expectedCells)
+                .mapToValue((writtenCell, expectedValue) -> {
+                    Optional<Integer> actualValue = storeToValidate.get(writtenCell.tableName(), writtenCell.cell());
+                    if (!actualValue.map(expectedValue::equals).orElse(false)) {
+                        return Optional.of(MismatchedValue.of(actualValue, Optional.of(expectedValue)));
+                    }
+                    return Optional.<MismatchedValue>empty();
+                })
+                .removeValues(Optional::isEmpty)
+                .mapValues(Optional::get)
+                .toMap();
+    }
+
+    private Map<TableWorkloadCell, Integer> findCellsThatAreNotDeleted(
+            Set<TableWorkloadCell> cells, ReadableTransactionStore storeToValidate) {
+        return StreamEx.of(cells)
+                .mapToEntry(cell -> storeToValidate.get(cell.tableName(), cell.cell()))
+                .removeValues(Optional::isEmpty)
+                .mapValues(Optional::get)
+                .toMap();
+    }
+
+    private static void logMismatchingCells(Map<TableWorkloadCell, MismatchedValue> mismatchingCells) {
+        mismatchingCells.forEach((cell, mismatchedValue) -> log.error(
```

Comment:
I really dislike log lines with very large parameters -- they can be hard to read, and the format of the map here isn't the friendliest. Realistically we don't expect this to fire, except rarely, so this should have minimal amount of logs. Of course, if this changes, then let's go ahead and just log a single-line with the map.

Another added benefit of having single vs multiple log lines, it's much easier to assess the impact of event by looking at frequency.